subprojects/maven/src/main/java/org/gradle/api/publish/maven/tasks/GenerateMavenPom.java

```diff
@@ -40,10 +41,11 @@
  */
 public class GenerateMavenPom extends DefaultTask {
 
-    private MavenPom pom;
+    private transient MavenPom pom;
```

Comment:
It seems we're relying on `transient` more and more. It's not a great contract, though, as it's not particularly descriptive and also not particularly reliable, in the sense that we don't do anything to prevent you from accessing the field after the "line" although it's being used to mean "I promise I won't use this after the line nor will anyone who attaches an action to this task" which no-one can really promise.

I wonder whether we should formalise this pattern a bit to make it more robust and clearer to use. We basically have some data that is an input to the calculation of the actual or "effective" parameters for the task action, and we might give people some way to structure their data or code to reflect this. For example using some structure (a `Property` subtype say) where I can set/query/mutate the inputs of the calculation prior to the line but not after the line. And vice versa for the output of the calculation. Or another strategy might be some declarative way to separate the DSL "shell" from the action and its effective parameters, where the shell carries the configuration time model and can be accessed and mutated prior to the line, then we serialize the action and its effective parameters, but not the shell. The action would not be allowed to access the shell but only its parameters. Or perhaps an annotation on the field to indicate that it is an input into the effective parameters calculation and cannot be accessed after the line. The decoration could then prevent access to the field.

Anyway, this is only an observation of a potential anti-pattern creeping in and something we can improve later. We should, however, probably add some safe guards to the public getters and other method that use these transient fields, to give users a better error message if they happen to add a task action that uses the getters, rather than a mysterious NPE. Same for the other tasks where we've used `transient` fields to work around the missing expressiveness.

Reply:
Thanks for the review, @adammurdoch. I think your observation makes a lot of sense and I agree some more formalisation of this pattern could prove beneficial.

For now, I introduced the lightweight `Transient<T>` and `Transient.Var<T>` types to at least provide better diagnostics whenever these fields are used after the "line".